### ISSUE TITLE
Version of duk_json_encode() allowing custom indent

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#408]
Current implementation of `duk_json_encode()` uses default encoding settings, putting the entire JSON literal on one long line.  Setting indent requires using the valstack API to call JSON.stringify() directly from C, which is not optimal since if a script reassigns the `JSON` object an unwanted error will be thrown.

What I suggest here is a version of `duk_json_encode()` allowing custom indent to be specified, useful for programs such as my Cell compiler, which generates source maps and manifests in JSON, where I need the output to be human-readable.

**Comment by svaarala** [comment]
I agree it'd be nice. What would you suggest for an API -- considering that there'd certainly be further options other than indent down the line?

I didn't add options to the API for this reason originally: a plain flags field won't be enough and adding individual options one by one as they come along would be awkward for compatibility.

**Comment by fatcerberus** [comment]
`JSON.stringify()` itself has only two options: indent and reviver.  I'm not certain a reviver function is useful from the native side, although it would be easy to have the new API accept a function pointer for it, or perhaps a boolean flag which, if true, uses a function pushed on the valstack.

For indent, the challenge is that in JS, it is either an arbitrary string or an integer specifying the number of spaces to indent per level.  This would require 2 functions at the least, which is a bit excessive for such a basic primitive.

One option that would be fairly elegant might be to get encoding options from the valstack, and accept a bitmask specifying which ones are present, similar to how `duk_def_prop()` is set up.

**Comment by svaarala** [comment]
I've come to pretty similar conclusions, I guess the best options would be:

- A flags + variable value stack approach similar to `duk_def_prop()`. This is quite flexible but not easy to read because the value stack behavior depends on flags.
- Provide an optional value stack index pointing to an object encoding control arguments like: `duk_json_encode(ctx, -3 /*obj_idx*/, -7 /*opt_idx*/)`. The value at "opt_idx" (if != DUK_INVALID_INDEX) would be something like: `{ indent: 4, reviver: myReviver, customOptionX: true }`.

The latter approach would also work well for the Ecmascript API, so that you could provide a table of arguments to Duktape.enc(), like:

    Duktape.enc('jx', myVal, {
        indent: 4, asciiOnly: true, allowCycles: true
    });  // arg examples are bogus :)

The downside of this is the relative cost of creating and extracting the arguments. The arguments can often be shared/reused for a specific call site.

This would be a good place to add a "multi-get" primitive to extract multiple values using a format string, like:

    indent=0; ascii=1;  /* defaults */
    duk_get_prop_multi(ctx, -3, "indent:%d, asciiOnly:%b", &indent, &ascii);

**Comment by fatcerberus** [comment]
I like that multi-get idea, it reminds me of C `scanf()`.

**Comment by svaarala** [comment]
Jansson (JSON library) has a similar "unpack" syntax and it's quite convenient.

A multi-get might even be footprint neutral: if there are just a few significant internal call sites it will most likely offset having it included. It'd probably be best to add it first internally and see what works best.

**Comment by svaarala** [comment]
@fatcerberus What do you think about argument object vs. flags + variable value stack?

**Comment by fatcerberus** [comment]
The argument object is more JS idiomatic, but my instinctive preference is for flags + valstack because constructing objects on the stack is one of the more annoying things I've had to do with Duktape.  Push object + 2 or more lines per property vs. 1 line per property.  The resulting code also tends to be difficult to follow, especially in case of a complex object.

Plus the stack approach seems more C-idiomatic.

Admittedly with flags it can get confusing what order to push the variables, but that's what the docs are for, and I'll take that over ~10 lines of code to set up a function call. :)

To see what I mean, check out this bit in minisphere to compile a CoffeeScript file:
https://github.com/fatcerberus/minisphere/blob/master/src/script.c#L80-L93

**Comment by svaarala** [comment]
Right. But what if there was also a `duk_put_prop_multi()`?

```
duk_push_object(ctx);
duk_put_prop_multi(ctx, -1, "indent:%d allowCyclic:%b", 4, 1);
```

**Comment by fatcerberus** [comment]
Hm, that would definitely make it less onerous to work with objects in native.  I think at that point I'd be neutral as to which solution is used. :)

**Comment by svaarala** [comment]
Or with a bit more concreteness:

```
duk_push_object(ctx);
duk_put_prop_multi(ctx, -1, "indent:%d allowCyclic:%b", 4, 1);
duk_json_encode(ctx, idx_obj, -1 /*opt_idx*/);
```

Vs:

```
duk_push_int(ctx, 4);
duk_json_encode(ctx, idx_obj, DUK_ENC_ALLOW_CYCLIC | DUK_ENC_HAVE_INDENT);
```

**Comment by svaarala** [comment]
Added a separate pull for the multi get/put, #409.

**Comment by fatcerberus** [comment]
Okay, yes, the multi-put version is definitely more readable than the flags variant.  So yeah, I'd concur to go with that route.
